src/esmac_diags/preprocessing/prep_ENA_E3SM.py

```diff
@@ -1,993 +1,1150 @@
+    matched_vlist = list(set(av_vars).intersection(req_vlist))
+    
+    if len(matched_vlist) == len(req_vlist):
+        cdnc_col = e3smdata['CDNUMC'+'_'+E3SMdomain_range].load()
+        cdnc_col = cdnc_col[:,x_idx]
+        cdnc_mean = cdnc_col/np.sum(weight,axis=1)
+        cdnc_mean[cdnc_mean >2e9] = np.nan
+        cdnc_mean = xr.DataArray(data=cdnc_mean,  dims=["time"],
+            coords=dict(time=(["time"], e3smtime)),
+            attrs=dict(long_name="mean cloud water number concentration",units="#/m3"),)
+    else:
+        cdnc_mean = xr.DataArray(np.zeros(len(e3smtime))*np.nan,attrs={'units':'dummy_unit','long_name':'Dummy'})
+    
+    # cloud droplet number concentration retrieved like Ndrop and Bennartz 2007
+    req_vlist = ['TGCLDLWP']
+    req_vlist = ["{}_{}".format(i,E3SMdomain_range) for i in req_vlist]
+    matched_vlist = list(set(av_vars).intersection(req_vlist))
+    
+    if len(matched_vlist) == len(req_vlist):
+        lwp = e3smdata['TGCLDLWP'+'_'+E3SMdomain_range][:,x_idx].data
+        e3sm_cloud_depth[z_cldtop>5000] = np.nan  # remove deep clouds with cloud top >5km
+        T_cldtop[z_cldtop>5000] = np.nan  # remove deep clouds with cloud top >5km
+        nd_arm = calc_cdnc_ARM(lwp, cod_m, e3sm_cloud_depth)
+        nd_sat = calc_cdnc_VISST(lwp, T_cldtop, cod_m, adiabaticity=0.8)
+        cdnc_arm = xr.DataArray(data=nd_arm*1e6,  dims=["time"],
+            coords=dict(time=(["time"], e3smtime)),
+            attrs=dict(long_name="mean cloud water number concentration",units="#/m3",\
+                        description='Retrieved using ARM Ndrop algorithm'),)
+        cdnc_sat = xr.DataArray(data=nd_sat*1e6,  dims=["time"],
+            coords=dict(time=(["time"], e3smtime)),
+            attrs=dict(long_name="mean cloud water number concentration",units="#/m3",\
+                        description='Retrieved using Bennartz(2007) algorithm, also used for VISST data'),)
+    else:
+        cdnc_arm = xr.DataArray(np.zeros(len(e3smtime))*np.nan,attrs={'units':'dummy_unit','long_name':'Dummy'})
+        cdnc_sat = xr.DataArray(np.zeros(len(e3smtime))*np.nan,attrs={'units':'dummy_unit','long_name':'Dummy'})
+    
+    # all other 2D (surface and vertical integrated) variables
+    variable2d_names = ['AODABS', 'AODALL', 'CDNUMC', 'CLDHGH', 'CLDMED', 'CLDLOW', 'CLDTOT', 
+                        'FLDS', 'FLNS', 'FLNT', 'FLUT', 'FSDS', 'FSNS', 'FSNT', 'FSUTOA', 'SOLIN', 
+                        'LHFLX', 'SHFLX', 'LWCF', 'SWCF', "TGCLDLWP", "TGCLDIWP", 
+                        'IWPMODIS', 'LWPMODIS', 'REFFCLWMODIS', 'TAUIMODIS', 'TAUTMODIS', 'TAUWMODIS', 
+                        #'MEANPTOP_ISCCP', 'MEANCLDALB_ISCCP', 'MEANTAU_ISCCP',
+                        'PBLH', 'PRECT', 'PRECL', 'PRECC', 'PS', 'TREFHT', ]
+    
+    # getting column and levels
+    len_ncol = len(e3smdata['ncol_'+E3SMdomain_range])
+    len_lev = len(e3smdata['lev'])
```

Comment:
same comments as above for all other files.

Reply:
Moved these at the top of the function.